vendor/blockchain_contracts/src/ethereum/rfc003/erc20_htlc.rs

```diff
@@ -0,0 +1,123 @@
+use crate::offset_parameter::{apply_offsets, OffsetParameter};
+use hex_literal::hex;
+use std::ops::Range;
+use web3::types::{Address, Bytes, U256};
+
+pub const SECRET_HASH_RANGE: Range<usize> = 53..85;
+pub const EXPIRY_RANGE: Range<usize> = 102..106;
+pub const REDEEM_IDENTITY_RANGE: Range<usize> = 157..177;
+pub const REFUND_IDENTITY_RANGE: Range<usize> = 224..244;
+pub const TOKEN_QUANTITY_RANGE: Range<usize> = 261..293;
+pub const TOKEN_CONTRACT_RANGE: Range<usize> = 307..327;
+
+const CONTRACT_TEMPLATE: [u8;339] = hex!("61014461000f6000396101446000f3361561005457602036141561006057602060006000376020602160206000600060026048f17f000000000000000000000000000000000000000000000000000000000000000060215114166100665760006000f35b426300000000106100a9575b60006000f35b7fb8cac300e37f03ad332e581dea21b2f0b84eaaadc184a295fef71e81f44a741360206000a17300000000000000000000000000000000000000006020526100ec565b7f5d26862916391bf49478b2f5103b0720a842b45ef145a268f2cd1fb2aed5517860006000a17300000000000000000000000000000000000000006020526100ec565b63a9059cbb6000527f0000000000000000000000000000000000000000000000000000000000000064604052602060606044601c6000730000000000000000000000000000000000000000620186a05a03f150602051ff");
+
+#[derive(Debug, Clone)]
+pub struct Erc20Htlc(Vec<u8>);
+
+impl From<Erc20Htlc> for Vec<u8> {
+    fn from(htlc: Erc20Htlc) -> Self {
+        htlc.0
+    }
+}
+
+impl Erc20Htlc {
+    pub fn new(
+        expiry: u32,
+        refund_identity: Address,
+        redeem_identity: Address,
+        secret_hash: [u8; 32],
+        token_contract_address: Address,
+        token_quantity: U256,
+    ) -> Self {
+        let offsets = vec![
+            OffsetParameter::new(expiry, EXPIRY_RANGE).expect("always 4 bytes"),
```

Comment:
:wrench: These expects are annoying. I think we can redesign this bit so that we give each item a mutable slice that they have to place themselves in.

e.g. `fn write_to_evm_contract(&self, slice: &mut [u8])`